colossalai/fx/passes/utils.py

```diff
@@ -25,3 +27,134 @@ def get_comm_size(prev_partition, next_partition):
                 comm_size += n.meta['tensor_meta'].numel
                 visited_nodes.add(n)
     return comm_size
+
+
+def get_leaf(graph: Graph):
+    """
+    Given a graph, return leaf nodes of this graph.
+
+    Note: This method will get the leaf nodes of given graph excluding ``output`` nodes.
+    """
+    input_nodes: Dict[Node, None] = {}
+    for node in graph.nodes:
+        if node.op == 'output':
+            map_arg(node.args, lambda n: input_nodes.setdefault(n))
+            map_arg(node.kwargs, lambda n: input_nodes.setdefault(n))
+    placeholder_nodes = []
+    for node in input_nodes.keys():
+        if node.op == 'placeholder':
+            placeholder_nodes.append(node)
+    for node in placeholder_nodes:
+        input_nodes.pop(node)
+    return list(input_nodes.keys())
+
+
+def is_leaf(graph: Graph, node: Node):
+    return node in get_leaf(graph)
+
+
+def get_top(graph: Graph):
+    """
+    Given a graph, return top nodes of this graph.
+
+    Note: This method will get the top nodes of given graph excluding ``placeholder`` nodes.
+    """
```

Comment:
Top nodes are quite confusing, will `get_placeholder` be better?

Reply:
This method is used to get the root node after removing 'placeholder' and 'root' node from the origin fx graph.

Reply:
Ok, can you add this definition of top node in the docstring as it is not a common terminology?

Reply:
Sure